File: ios/Runner/JSFlutter/MXJSFlutterEngine.m
//
//  MXJSFlutterEngine.m
//  Runner
//
//  Created by soapyang on 2018/12/22.
//  Copyright © 2018 The Chromium Authors. All rights reserved.
//

#import "MXJSFlutterEngine.h"
#import "MXJSFlutterApp.h"
#import "JSModule.h"
#import "MXJSFlutterDefines.h"
#import <Flutter/Flutter.h>
#import "AppDelegate.h"
#import "MXJSEngine.h"

@interface MXJSFlutterEngine ()
{
//    BOOL _flutterEngineIsDidRender;
}

@property (nonatomic, strong) NSString *rootPath;
@property (nonatomic, strong) MXJSFlutterApp *currentApp;

@property (nonatomic, strong) FlutterMethodChannel *basicChannel;
@property (nonatomic, strong) NSMutableArray<FlutterMethodCall*> *callFlutterQueue;

@end

@implementation MXJSFlutterEngine

- (instancetype)initJSAppName:(NSString*)jsAppName
{
    if (self = [super init])
    {
//        _flutterEngineIsDidRender = NO;
        _jsAppName = jsAppName;
        
        [self setup];
    }
    return self;
}

- (void)setup
{
    [self unsetup];
    
    self.rootPath = [JSFLUTTER_SRC_BASE_DIR stringByAppendingPathComponent:JSFLUTTER_SRC_DIR];
    self.callFlutterQueue = [NSMutableArray arrayWithCapacity:2];
//    __weak MXJSFlutterViewController *weakSelf = self;
//    
//    [self setFlutterViewDidRenderCallback:^{
//        _flutterEngineIsDidRender = YES;
//        
//        for (FlutterMethodCall *call in weakSelf.callFlutterQueue) {
//            [weakSelf.basicChannel invokeMethod:call.method arguments:call.arguments];
//        }
//    }];
    
    AppDelegate *delegate = ((AppDelegate *)UIApplication.sharedApplication.delegate);
    self.flutterEngine = delegate.flutterEngine;
    
    [self setupChannel];
    
    if (self.jsAppName && self.jsAppName.length > 0)
    {
        [self runJSApp:self.jsAppName];
    }
}

- (void)setupChannel
{
    
    self.basicChannel = [FlutterMethodChannel
                         methodChannelWithName:@"js_flutter.flutter_main_channel"
                         binaryMessenger:self.flutterEngine.binaryMessenger];
    
    __weak MXJSFlutterEngine *weakSelf = self;
    [self.basicChannel setMethodCallHandler:^(FlutterMethodCall * _Nonnull call, FlutterResult  _Nonnull result) {
        __strong MXJSFlutterEngine *strongSelf = weakSelf;
        if (!strongSelf) {
            return;
        }
        
        if ([call.method isEqualToString:@"callNativeRunJSApp"]) {
            [strongSelf callNativeRunJSApp:call.arguments];
        } else if ([call.method isEqualToString:@"callJsCallbackFunction"]) {
            [strongSelf callJsCallBackFunction:call.arguments];
        }
    }];
    
    //Test
    FlutterMethodChannel* batteryChannel = [FlutterMethodChannel methodChannelWithName:@"samples.flutter.io/battery" binaryMessenger:self.flutterEngine.binaryMessenger];
    [batteryChannel setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result) {
        if ([call.method isEqualToString:@"getPlatformVersion"]) {
            result(@"samples.flutter.io/battery test string");
        } else {
            result(@(404));
        }
    }];
}

- (void)runJSApp:(NSString*)appName
{
    [self runApp:appName pageName:nil];
}

- (void)callJsCallBackFunction:(id)arguments
{
    NSDictionary *argsMap = arguments;
    NSString *callbackId = argsMap[@"callbackId"];
    NSString *param = argsMap[@"param"];
    
    [self.jsEngine callJSCallbackFunction:callbackId param:param];
}

//MARK: - flutter -> Native
//由Flutter 代码启动JSApp。 可以用在先显示Dart页面，然后路由调转到JS页面
//启动JSApp之后，执行JS代码，JS代码可以主动调用Flutter显示自己的页面，也能接受Flutter的指令，显示对应页面
- (void)callNativeRunJSApp:(id)arguments
{
    NSDictionary *argsMap = arguments;
    NSString *jsAppName = argsMap[@"jsAppName"];
    NSString *pageName = argsMap[@"pageName"];
    
    [self runApp:jsAppName pageName:pageName];
}

//MARK: - native -> flutter
//--------------------------------------------
//调用Flutter切换到JSWidget，显示JS渲染的内容
//MARK: - native -> flutter
//--------------------------------------------
- (void)callFlutterReloadAppWithJSWidgetData:(NSString*)widgetData
{
    [self callFlutterReloadAppWithRouteName:@"MXJSWidget" widgetData:widgetData];
}

- (void)callFlutterReloadAppWithRouteName:(NSString*)routeName widgetData:(NSString*)widgetData
{
    routeName = routeName?routeName:@"";
    widgetData = widgetData?widgetData:@"";
    
    NSMutableDictionary *arguments = [[NSMutableDictionary alloc] init];
    if (routeName)
    {
        [arguments setObject:routeName forKey:@"routeName"];
    }
    if (widgetData)
    {
        [arguments setObject:widgetData forKey:@"widgetData"];
    }
    FlutterMethodCall* call  = [FlutterMethodCall methodCallWithMethodName:@"reloadApp" arguments:arguments];
    
//    if (!_flutterEngineIsDidRender) {
//
//        [self.callFlutterQueue addObject:call];
//        return;
//    }
    
    [self.basicChannel invokeMethod:call.method arguments:call.arguments];
}

- (void)callFlutterMethodChannelInvoke:(NSString*)channelName methodName:(NSString*)methodName params:(NSDictionary *)params callback:(void(^)(id _Nullable result))callback
{
    NSMutableDictionary *arguments = [[NSMutableDictionary alloc] init];
    if (channelName)
    {
        [arguments setObject:channelName forKey:@"channelName"];
    }
    if (methodName)
    {
        [arguments setObject:methodName forKey:@"methodName"];
    }
    if (params)
    {
        [arguments setObject:params forKey:@"params"];
    }
    FlutterMethodCall* call = [FlutterMethodCall methodCallWithMethodName:@"mxflutterBridgeMethodChannelInvoke" arguments:arguments];
//    if (!_flutterEngineIsDidRender) {
//        [self.callFlutterQueue addObject:call];
//        return;
//    }
    
    [self.basicChannel invokeMethod:call.method arguments:call.arguments result:^(id  _Nullable result) {
        if (callback) {
            callback(result);
        }
    }];
}

- (void)callFlutterEventChannelReceiveBroadcastStreamListenInvoke:(NSString*)channelName
                                                      streamParam:(NSString *)streamParam
                                                         onDataId:(NSString *)onDataId
                                                        onErrorId:(NSString *)onErrorId
                                                         onDoneId:(NSString *)onDoneId
                                                    cancelOnError:(NSNumber *)cancelOnError
                            
{
    if ([streamParam isEqualToString:@"null"]) {
        streamParam = nil;
    }
    
    NSMutableDictionary *arguments = [[NSMutableDictionary alloc] init];
    if (channelName)
    {
        [arguments setObject:channelName forKey:@"channelName"];
    }
    if (streamParam)
    {
        [arguments setObject:streamParam forKey:@"streamParam"];
    }
    if (onDataId)
    {
        [arguments setObject:onDataId forKey:@"onDataId"];
    }
    if (onErrorId)
    {
        [arguments setObject:onErrorId forKey:@"onErrorId"];
    }
    if (onErrorId)
    {
        [arguments setObject:onErrorId forKey:@"onErrorId"];
    }
    if (cancelOnError)
    {
        [arguments setObject:cancelOnError forKey:@"cancelOnError"];
    }
    FlutterMethodCall* call = [FlutterMethodCall methodCallWithMethodName:@"mxflutterBridgeEventChannelReceiveBroadcastStreamListenInvoke"
                                                                arguments:arguments];
//    if (!_flutterEngineIsDidRender) {
//        [self.callFlutterQueue addObject:call];
//        return;
//    }
    
    [self.basicChannel invokeMethod:call.method arguments:call.arguments];
}


- (void)unsetup {
    if (self.currentApp) {
        [self.currentApp exitApp];
        self.currentApp = nil;
    }
}

- (void)runApp:(NSString*)appName pageName:(NSString*)pageName
{
    [self setup];
    [JSModule clearModuleCache];
    NSString *appRootPath = [self.rootPath stringByAppendingPathComponent:appName];
    self.currentApp  = [[MXJSFlutterApp alloc] initWithAppName:appName engine:self appRootPath:appRootPath];
    
    [self.currentApp runAppWithPageName:pageName];
}

- (BOOL)showPage:(NSString*)pageName
{
    return YES;
}

@end

File: mxflutter/android/src/main/java/com/mojitox/mxflutter/framework/MXJSFlutterEngine.java
//  MXFlutterFramework
//  Copyright 2019 The MXFlutter Authors. All rights reserved.
//
//  Use of this source code is governed by a MIT-style license that can be
//  found in the LICENSE file.

package com.mojitox.mxflutter.framework;

import android.text.TextUtils;

import com.mojitox.mxflutter.MXFlutterPlugin;
import com.mojitox.mxflutter.framework.utils.FileUtils;
import com.mojitox.mxflutter.framework.utils.LogUtilsKt;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.flutter.Log;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.BasicMessageChannel;
import io.flutter.plugin.common.StringCodec;

public class MXJSFlutterEngine {

    public static final String TAG = "MXJSFlutterEngine";

    public String mJsFrameworkPath;
    public String mCurrentJSAppPath;
    public ArrayList<String> mJsAppSearchPathList;

    private MXJSFlutterApp currentApp;

    private MXFlutterPlugin mContext;
    private BinaryMessenger mFlutterEngine;

    private MXJSEngine mJsEngine;

    //Flutter通道
    private static final String FLUTTER_METHED_CHANNEL_NAME = "js_flutter.flutter_main_channel";
    private static final String FLUTTER_COMMON_BASIC_CHANNEL_NAME = "mxflutter.mxflutter_common_basic_channel";
    MethodChannel jsFlutterAppChannel;
    BasicMessageChannel<String> jsFlutterCommonBasicChannel;

    private boolean isFlutterEngineIsDidRender;
    private ArrayList<MethodCall> callFlutterQueue;

    public MXJSFlutterEngine(MXFlutterPlugin context, BinaryMessenger flutterEngine) {
        this.mContext = context;
        this.mFlutterEngine = flutterEngine;
        setup();
    }

    public void setJsEngine(MXJSEngine jsEngine) {
        mJsEngine = jsEngine;
    }

    public void setup() {
        setupChannel();
//        callFlutterQueue = new ArrayList<>(1);
//        isFlutterEngineIsDidRender = true;
//        for (MethodCall call : callFlutterQueue) {
//            jsFlutterAppChannel.invokeMethod(call.method, call.arguments);
//        }
    }

    public void setupChannel() {
        jsFlutterAppChannel = new MethodChannel(mFlutterEngine, FLUTTER_METHED_CHANNEL_NAME);
        jsFlutterAppChannel.setMethodCallHandler(new MethodChannel.MethodCallHandler() {
            @Override
            public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {
                if (methodCall.method.equals("callNativeRunJSApp")) {
                    String jsAppPath = methodCall.argument("jsAppPath");
                    String jsAppAssetsKey = methodCall.argument("jsAppAssetsKey");
                    List<String> jsAppSearchPathList = (List<String>) methodCall.argument("jsAppSearchPathList");
                    List<String> jsAppSearchPathWithAssetsKeyList = (List<String>) methodCall.argument("jsAppSearchPathWithAssetsKeyList");
                    runApp(jsAppPath, jsAppAssetsKey, jsAppSearchPathList, jsAppSearchPathWithAssetsKeyList);
                    result.success("success");
                } else if (methodCall.method.equals("callJsCallbackFunction")) {
                    String jsAppName = methodCall.argument("callbackId");
                    Map param = methodCall.argument("param");
                    mJsEngine.callJSCallbackFunction(jsAppName, param);
                    result.success("success");
                } else if (methodCall.method.equals("mxLog")) {
                    Log.i(TAG, methodCall.arguments.toString());
                }
            }
        });

        jsFlutterCommonBasicChannel = new BasicMessageChannel<>(mFlutterEngine, FLUTTER_COMMON_BASIC_CHANNEL_NAME, StringCodec.INSTANCE);
        jsFlutterCommonBasicChannel.setMessageHandler(new BasicMessageChannel.MessageHandler<String>() {
            @Override
            public void onMessage(String message, BasicMessageChannel.Reply<String> reply) {
                mJsEngine.jsExecutor.invokeJSValueWithString(MXJSExecutor.runtime, "mxfJSBridgeInvokeJSCommonChannel", message, new MXJSExecutor.InvokeJSValueCallback() {
                    @Override
                    public void onSuccess(Object value) {
                        reply.reply(value.toString());
                    }

                    @Override
                    public void onError(Error error) {

                    }
                });
            }
        });
    }

    public void destroy() {
        if (currentApp != null) {
            currentApp.close();
        }
    }

    public boolean showPage(String pageName) {
        return true;
    }

    public void runApp(String jsAppPath, String jsAppAssetsKey, List<String> jsAppSearchPathList, List<String> jsAppSearchPathWithAssetsKeyList) {
        Set<String> searchList = new HashSet<String>();

        if (!TextUtils.isEmpty(mCurrentJSAppPath)) {
            LogUtilsKt.MXJSFlutterLog("MXJSFlutterEngine Native工程，设置了currentJSAppPath，RunJSApp使用Native设置的路径 appPath:%s", mCurrentJSAppPath);
            jsAppPath = mCurrentJSAppPath;
            jsAppSearchPathList = mJsAppSearchPathList;
        } else {
            if (TextUtils.isEmpty(jsAppPath) && !TextUtils.isEmpty(jsAppAssetsKey)) {
                jsAppPath = mContext.mFlutterPluginBinding.getFlutterAssets().getAssetFilePathByName(jsAppAssetsKey);
            }
            if (TextUtils.isEmpty(jsAppPath)) {
                LogUtilsKt.MXJSFlutterLog("jsAppPath == null", "");
                return;
            }
        }
        if (jsAppSearchPathList != null && jsAppSearchPathList.size() > 0)
            searchList.addAll(jsAppSearchPathList);
        if (jsAppSearchPathWithAssetsKeyList != null && jsAppSearchPathWithAssetsKeyList.size() > 0) {
            for (String searchPathAssetKey : jsAppSearchPathWithAssetsKeyList) {
                String path;
                if(FileUtils.isCopiedFileFromAssets(mContext.mFlutterPluginBinding.getApplicationContext())){
                    path = searchPathAssetKey;
                }else {
                    path = mContext.mFlutterPluginBinding.getFlutterAssets().getAssetFilePathByName(searchPathAssetKey);
                }
                if (!TextUtils.isEmpty(path) && !searchList.contains(path)) {
                    searchList.add(path);
                }
            }
        }
        jsAppSearchPathList = new ArrayList<>(searchList);

        if (currentApp != null) {
            currentApp.close();
            currentApp = null;
        }

        currentApp = new MXJSFlutterApp();
        currentApp.initWithAppName(mContext, jsAppPath, jsAppPath, jsAppSearchPathList, this);
        currentApp.runAppWithPageName();
    }

    public void callFlutterReloadAppWithJSWidgetData(String widgetData) {
        callFlutterReloadAppWithRouteName("MXJSWidget", widgetData);
    }

    @SuppressWarnings("unchecked")
    public void callFlutterReloadAppWithRouteName(String routeName, String widgetData) {
        FileUtils.assertUiThread();

        if (routeName == null || widgetData == null) {
            return;
        }

        Map arg = new HashMap();
        arg.put("routeName", routeName);
        arg.put("widgetData", widgetData);
        MethodCall call = new MethodCall("reloadApp", arg);
//      if (isFlutterEngineIsDidRender) {
//          callFlutterQueue.add(call);
//          return;
//      }
        jsFlutterAppChannel.invokeMethod(call.method, call.arguments);
    }

    //顶层通用通道
    @SuppressWarnings("unchecked")
    public void invokeFlutterCommonChannel(String callJSONStr, BasicMessageChannel.Reply<String> Reply) {
        FileUtils.assertUiThread();
        jsFlutterCommonBasicChannel.send(callJSONStr, Reply);
    }

    @SuppressWarnings("unchecked")
    public void callFlutterMethodChannelInvoke(String channelName, String methodName, Map params, MethodChannel.Result callback) {
        FileUtils.assertUiThread();

        Map arg = new HashMap();
        arg.put("channelName", channelName);
        arg.put("params", params);
        arg.put("methodName", methodName);

        MethodCall call = new MethodCall("mxflutterBridgeMethodChannelInvoke", arg);
        jsFlutterAppChannel.invokeMethod(call.method, call.arguments, callback);
    }

    @SuppressWarnings("unchecked")
    public void callFlutterEventChannelReceiveBroadcastStreamListenInvoke(String channelName, String streamParam, String onDataId, String onErrorId, String onDoneId, boolean cancelOnError) {
        FileUtils.assertUiThread();

        Map arg = new HashMap();
        arg.put("channelName", channelName);
        arg.put("streamParam", streamParam);
        arg.put("onDataId", onDataId);
        arg.put("onErrorId", onErrorId);
        arg.put("onDoneId", onDoneId);
        arg.put("cancelOnError", cancelOnError);

        MethodCall call = new MethodCall("mxflutterBridgeEventChannelReceiveBroadcastStreamListenInvoke", arg);
        jsFlutterAppChannel.invokeMethod(call.method, call.arguments);
    }
}
